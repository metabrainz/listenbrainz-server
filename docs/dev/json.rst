.. _json-doc:

JSON Documentation
==================

.. note:: Do not submit copyrighted information in these fields!

Submission JSON
---------------

To submit a listen via our API (see: :doc:`api`), ``POST`` a JSON document to
the ``submit-listens`` endpoint. Submit one of three types JSON documents:

- ``single``: Submit single listen

   - Indicates user just finished listening to track

   - ``payload`` should contain information about *exactly one* track

- ``playing_now``: Submit ``playing_now`` notification

   - Indicates that user just began listening to track

   - ``payload`` should contain information about *exactly one* track

   - Submitting ``playing_now`` documents is optional

   - Timestamp must be omitted from a ``playing_now`` submission.

- ``import``: Submit previously saved listens

   - ``payload`` should contain information about *at least one* track

   - submitting multiple listens in one go is allowed, but the complete JSON
     document may not exceed :data:`~webserver.views.api.MAX_LISTEN_SIZE` bytes
     in size

The ``listen_type`` element defines different types of submissions. The element
is placed at the top-most level of the JSON document. The only other required
element is the ``payload`` element. This provides an array of listens â€“ the
payload may be one or more listens (as designated by ``listen_type``)::

    {
      "listen_type": "single",
      "payload": [
          --- listen data here ---
      ]
    }

A sample listen payload may look like::

    {
      "listened_at": 1443521965,
      "track_metadata": {
        "additional_info": {
          "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
          "artist_mbids": [
            "db92a151-1ac2-438b-bc43-b82e149ddd50"
          ],
          "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
          "tags": [ "you", "just", "got", "rick rolled!"]
        },
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
        "release_name": "Whenever you need somebody"
      }
    }

A complete submit listen JSON document may look like::

    {
      "listen_type": "single",
      "payload": [
        {
          "listened_at": 1443521965,
          "track_metadata": {
            "additional_info": {
              "listening_from": "spotify",
              "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
              "artist_mbids": [
                "db92a151-1ac2-438b-bc43-b82e149ddd50"
              ],
              "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
              "tags": [ "you", "just", "got", "rick rolled!"]
            },
            "artist_name": "Rick Astley",
            "track_name": "Never Gonna Give You Up",
            "release_name": "Whenever you need somebody"
          }
        }
      ]
    }


Fetching listen JSON
--------------------

The JSON documents returned from our API look like the following::

    {
      "payload": {
        "count": 25,
        "user_id": "-- the MusicBrainz ID of the user --",
        "listens": [
          "-- listen data here ---"
        ]
      }
    }

The number of listens in the document are returned by the top-level ``count``
element. The ``user_id`` element contains the MusicBrainz ID of the user whose listens are
being returned. The other element is the ``listens`` element. This is a list which contains
the listen JSON elements (described above).

The JSON document returned by the API endpoint for getting tracks being played right now
is the same as above, except that it also contains the ``payload/playing_now`` element as a
boolean set to True.


Payload JSON details
--------------------

A minimal payload must include
``track_metadata/artist_name`` and ``track_metadata/track_name`` elements::

    {
      "track_metadata": {
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
      }
    }

``artist_name`` and ``track_name`` elements must be simple strings.

The payload will also include the ``listened_at`` element which must be an integer
representing the Unix time when the track was listened to. The only exception to this
rule is when the listen is being played right now and has been retrieved from the
endpoint to get listens being played right now. The ``listened_at`` element will be
absent for such listens.

Add additional metadata you may have for a track to the ``additional_info``
element. Any additional information allows us to better correlate your listen
data to existing MusicBrainz-based data. If you have MusicBrainz IDs available,
submit them!

The following optional elements may also be included in the ``track_metadata`` element:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``release_name``        the name of the release this recording was played from.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

The following optional elements may also be included in the ``additional_info`` element. If you do not have
the data for any of the following fields, omit the key entirely:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``artist_mbids``        A list of MusicBrainz Artist IDs, one or more Artist IDs may be included here. If you have a complete MusicBrainz artist credit that contains multiple Artist IDs, include them all in this list.
``release_group_mbid``  A MusicBrainz Release Group ID of the release group this recording was played from.
``release_mbid``        A MusicBrainz Release ID of the release this recording was played from.
``recording_mbid``      A MusicBrainz Recording ID of the recording that was played.
``track_mbid``          A MusicBrainz Track ID associated with the recording that was played.
``work_mbids``          A list of MusicBrainz Work IDs that may be associated with this recording.
``tracknumber``         The tracknumber of the recording. This first recording on a release is tracknumber 1.
``isrc``                The ISRC code associated with the recording.
``spotify_id``          The Spotify track URL associated with this recording.  e.g.: http://open.spotify.com/track/1rrgWMXGCGHru5bIRxGFV0
``tags``                A list of user defined tags to be associated with this recording. These tags are similar to last.fm tags. For example, you have apply tags such as ``punk``, ``see-live``, ``smelly``. You may submit up to :data:`~webserver.views.api.MAX_TAGS_PER_LISTEN` tags and each tag may be up to :data:`~webserver.views.api.MAX_TAG_SIZE` characters large.
``listening_from``       The source of the listen, i.e the name of the client or service which submits the listen.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

At this point, we are not scrubbing any superflous elements that may be
submitted via the ``additional_info`` element. We're open to see how people
will make use of these unspecified fields and may decide to formally specify or
scrub elements in the future.
