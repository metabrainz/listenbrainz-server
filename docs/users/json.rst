.. _json-doc:

JSON Documentation
==================

.. note:: Do not submit copyrighted information in these fields!

Submission JSON
---------------

To submit a listen via our API (see: :doc:`api/core`), ``POST`` a JSON document to
the ``submit-listens`` endpoint. Submit one of three types JSON documents:

- ``single``: Submit single listen

   - Indicates user just finished listening to track

   - ``payload`` should contain information about *exactly one* track

- ``playing_now``: Submit ``playing_now`` notification

   - Indicates that user just began listening to track

   - ``payload`` should contain information about *exactly one* track

   - Submitting ``playing_now`` documents is optional

   - Timestamp must be omitted from a ``playing_now`` submission.

.. note::

    Playing Now listens are only stored temporarily. A playing now listen must be
    submitted again as a ``single`` or ``import`` for permanent storage.


- ``import``: Submit previously saved listens

   - ``payload`` should contain information about *at least one* track

   - submitting multiple listens in one go is allowed, but the complete JSON
     document may not exceed :data:`~webserver.views.api.MAX_LISTEN_SIZE` bytes
     in size

The ``listen_type`` element defines different types of submissions. The element
is placed at the top-most level of the JSON document. The only other required
element is the ``payload`` element. This provides an array of listens â€“ the
payload may be one or more listens (as designated by ``listen_type``):

.. code-block:: json

    {
      "listen_type": "single",
      "payload": [
          "--- listen data here ---"
      ]
    }

A sample listen payload may look like:

.. code-block:: json

    {
      "listened_at": 1443521965,
      "track_metadata": {
        "additional_info": {
          "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
          "artist_mbids": [
            "db92a151-1ac2-438b-bc43-b82e149ddd50"
          ],
          "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
          "tags": [ "you", "just", "got", "rick rolled!"]
        },
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
        "release_name": "Whenever you need somebody"
      }
    }

A complete submit listen JSON document may look like:

.. code-block:: json

    {
      "listen_type": "single",
      "payload": [
        {
          "listened_at": 1443521965,
          "track_metadata": {
            "additional_info": {
              "media_player": "Rhythmbox",
              "submission_client": "Rhythmbox ListenBrainz Plugin",
              "submission_client_version": "1.0",
              "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
              "artist_mbids": [
                "db92a151-1ac2-438b-bc43-b82e149ddd50"
              ],
              "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
              "tags": [ "you", "just", "got", "rick rolled!"]
            },
            "artist_name": "Rick Astley",
            "track_name": "Never Gonna Give You Up",
            "release_name": "Whenever you need somebody"
          }
        }
      ]
    }


Fetching listen JSON
--------------------

The JSON documents returned from our API look like the following:

.. code-block:: json

    {
      "payload": {
        "count": 25,
        "user_id": "-- the MusicBrainz ID of the user --",
        "listens": [
          "-- listen data here ---"
        ]
      }
    }

The number of listens in the document are returned by the top-level ``count``
element. The ``user_id`` element contains the MusicBrainz ID of the user whose listens are
being returned. The other element is the ``listens`` element. This is a list which contains
the listen JSON elements (described above).

The JSON document returned by the API endpoint for getting tracks being played right now
is the same as above, except that it also contains the ``payload/playing_now`` element as a
boolean set to True.


Payload JSON details
--------------------

A minimal payload must include
``track_metadata/artist_name`` and ``track_metadata/track_name`` elements:

.. code-block:: json

    {
      "track_metadata": {
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
      }
    }

``artist_name`` and ``track_name`` elements must be simple strings.

The payload should also include the ``listened_at`` element, which must be an integer
representing the Unix time when the track was listened to. The minimum accepted
value for this field is :data:`~webserver.views.api.LISTEN_MINIMUM_TS`.
playing_now requests should not have a ``listened_at`` field

Add additional metadata you may have for a track to the ``additional_info``
element. Any additional information allows us to better correlate your listen
data to existing MusicBrainz-based data. If you have MusicBrainz IDs available,
submit them!

The following optional elements may also be included in the ``track_metadata`` element:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``release_name``        the name of the release this recording was played from.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

The following optional elements may also be included in the ``additional_info`` element. If you do not have
the data for any of the following fields, omit the key entirely:

.. list-table:: Title
   :widths: 25 50
   :header-rows: 1

   * - element
     - description
   * - ``artist_mbids``
     - A list of MusicBrainz Artist IDs, one or more Artist IDs may be included here. If you have a complete MusicBrainz artist credit that contains multiple Artist IDs, include them all in this list.
   * - ``release_group_mbid``
     - A MusicBrainz Release Group ID of the release group this recording was played from.
   * - ``release_mbid``
     - A MusicBrainz Release ID of the release this recording was played from.
   * - ``recording_mbid``
     - A MusicBrainz Recording ID of the recording that was played.
   * - ``track_mbid``
     - A MusicBrainz Track ID associated with the recording that was played.
   * - ``work_mbids``
     - A list of MusicBrainz Work IDs that may be associated with this recording.
   * - ``tracknumber``
     - The tracknumber of the recording. This first recording on a release is tracknumber 1.
   * - ``isrc``
     - The ISRC code associated with the recording.
   * - ``spotify_id``
     - The Spotify track URL associated with this recording.  e.g.: http://open.spotify.com/track/1rrgWMXGCGHru5bIRxGFV0
   * - ``tags``
     - A list of user-defined folksonomy tags to be associated with this recording. For example, you have apply tags such as ``punk``, ``see-live``, ``smelly``. You may submit up to :data:`~webserver.views.api.MAX_TAGS_PER_LISTEN` tags and each tag may be up to :data:`~webserver.views.api.MAX_TAG_SIZE` characters large.
   * - ``media_player``
     - The name of the program being used to listen to music. Don't include a version number here.
   * - ``media_player_version``
     - The version of the program being used to listen to music.
   * - ``submission_client``
     - The name of the client that is being used to submit listens to ListenBrainz. If the media player has the ability to submit listens built-in then this value may be the same as ``media_player``. Don't include a version number here.
   * - ``submission_client_version``
     - The version of the submission client.
   * - ``music_service``
     - If the song being listened to comes from an online service, the canonical domain of this service (see below for more details).
   * - ``music_service_name``
     - If the song being listened to comes from an online service and you don't know the canonical domain, a name that represents the service.
   * - ``origin_url``
     - If the song of this listen comes from an online source, the URL to the place where it is available. This could be a spotify url (see ``spotify_id``), a YouTube video URL, a Soundcloud recording page URL, or the full URL to a public MP3 file. If there is a webpage for this song (e.g. Youtube page, Soundcloud page) **do not** try and resolve the URL to an actual audio resource.

.. note::

  **Music service names**

  The ``music_service`` field should be a domain name rather than a textual description or URL. This allows us to refer unambiguously to a service without worrying
  about capitalization or full/short names (such as the difference between "Internet Archive", "The Internet Archive" or "Archive").
  If we use this data on ListenBrainz, we will perform a mapping from the domain name to a canonical name. Below is an example of mappings that we currently support.
  If you are submitting from a service which doesn't appear in this list, you should determine a canonical domain from the domain of the service.
  Only if you cannot determine a domain for the service should you use the text-only ``music_service_name`` field.

  .. list-table:: Music services domain/name mapping
     :widths: 25 50
     :header-rows: 1

     * - domain
       - name
     * - ``spotify.com``
       - Spotify
     * - ``bandcamp.com``
       - Bandcamp
     * - ``youtube.com``
       - YouTube
     * - ``music.youtube.com``
       - YouTube Music
     * - ``deezer.com``
       - Deezer
     * - ``tidal.com``
       - TIDAL
     * - ``music.apple.com``
       - Apple Music
     * - ``archive.org``
       - Internet Archive
     * - ``soundcloud.com``
       - Soudcloud
     * - ``jamendo.com``
       - Jamendo Music
     * - ``play.google.com``
       - Google Play Music


Client Metadata examples
------------------------

Here are a few examples of how to fill in the ``media_player``, ``submission_client`` and ``music_service`` fields based on our
current recommendations.

BrainzPlayer on the ListenBrainz website playing a video from YouTube
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "BrainzPlayer",
            "music_service": "youtube.com",
            "origin_url": "https://www.youtube.com/watch?v=JKFBiaoFHoY",
            "submission_client": "BrainzPlayer"
        },
        "artist_name": "Mdou Moctar",
        "release_name": "Ilana (The Creator)",
        "track_name": "Inizgam"
    }
  }

BrainzPlayer on the ListenBrainz website playing a video from Spotify
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Note that even though the ``origin_url`` is ``https://open.spotify.com``, we set ``music_service``
to spotify.com (see above note).

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "BrainzPlayer",
            "music_service": "spotify.com",
            "origin_url": "https://open.spotify.com/track/5fEjp2F0Sqr9fMuLSaDqz0",
            "submission_client": "BrainzPlayer"
        },
        "artist_name": "Les Filles de Illighadad",
        "release_name": "Eghass Malan",
        "track_name": "Inssegh Inssegh"
    }
  }


Using Otter for Funkwhale on android, and submitting with Simple Scrobbler
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In this case, the media player and submission client are completely separate programs. Because music is being played
from a user's private collection and not a streaming service, don't include music_service or origin_url.

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "Otter",
            "media_player_version": "1.0.21",
            "submission_client": "Simple Scrobbler"
            "submission_client_version": "1.7.0"
        },
        "artist_name": "Les Filles de Illighadad",
        "release_name": "Eghass Malan",
        "track_name": "Inssegh Inssegh"
    }
  }


Rhythmbox player listening to Jamendo
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "Rhythmbox",
            "music_service": "jamendo.com",
            "music_service_name": "Jamendo Music"
            "origin_url": "https://www.jamendo.com/track/1466090/universal-funk",
            "submission_client": "Rhythmbox ListenBrainz Plugin"
        },
        "artist_name": "Duo Teslar",
        "track_name": "Universal Funk"
    }
  }

Listening to a recording from Bandcamp and submitting with the browser extension WebScrobbler
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Because playback happens in the browser, there is no specific ``media_player``.

.. code-block:: JSON

  {
	"track_metadata": {
		"additional_info": {
			"music_service": "bandcamp.com",
			"music_service_name": "Bandcamp",
			"submission_client": "WebScrobbler",
			"submission_client_version": "v2.48.0"
			"origin_url": "https://greencookierecords.bandcamp.com/track/shake",
		},
		"artist_name": "I Mitomani Beat",
		"release_name": "Fuori Dal Tempo",
		"track_name": "Shake",
	}
  }

At this point, we are not removing any other elements that may be
submitted via the ``additional_info`` element. We're open to see how people
will make use of these unspecified fields and may decide to formally specify or
scrub elements in the future.
