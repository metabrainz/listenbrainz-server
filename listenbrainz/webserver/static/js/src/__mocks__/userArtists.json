{
  "payload": {
    "artists": [
      {
        "artist_mbids": [],
        "artist_msid": "d340853d-7408-4a0d-89c2-6ff13e568815",
        "artist_name": "The Local train",
        "listen_count": 385
      },
      {
        "artist_mbids": [],
        "artist_msid": "ba64b195-01dd-4613-9534-bb87dc44cffb",
        "artist_name": "Lenka",
        "listen_count": 333
      },
      {
        "artist_mbids": [],
        "artist_msid": "6599e41e-390c-4855-a2ac-68ee798538b4",
        "artist_name": "Coldplay",
        "listen_count": 321
      },
      {
        "artist_mbids": [],
        "artist_msid": "fecda98e-0533-4b87-a9d1-c89b4521918b",
        "artist_name": "Imagine Dragons",
        "listen_count": 176
      },
      {
        "artist_mbids": [],
        "artist_msid": "9d5adabb-9775-4535-93eb-12b485aed8bf",
        "artist_name": "Maroon 5",
        "listen_count": 118
      },
      {
        "artist_mbids": [],
        "artist_msid": "7addbcac-ae39-4b4c-a956-53da336d68e8",
        "artist_name": "Ellie Goulding",
        "listen_count": 108
      },
      {
        "artist_mbids": [],
        "artist_msid": "67f4ae2e-260e-4020-b957-f2df5b719501",
        "artist_name": "OneRepublic",
        "listen_count": 65
      },
      {
        "artist_mbids": [],
        "artist_msid": "2b0646af-f3f0-4a5b-b629-6c31301c1c29",
        "artist_name": "The Weeknd",
        "listen_count": 52
      },
      {
        "artist_mbids": [],
        "artist_msid": "0daecabb-f76d-4914-8726-82cefbecca52",
        "artist_name": "The Chainsmokers",
        "listen_count": 44
      },
      {
        "artist_mbids": [],
        "artist_msid": "3b155259-b29e-4515-aa62-cb0b917f4cfd",
        "artist_name": "The Fray",
        "listen_count": 41
      }
    ],
    "count": 10,
    "last_updated": 1588494361,
    "offset": 0,
    "range": "all_time",
    "total_artist_count": 175,
    "user_id": "ishaanshah"
  }
}
