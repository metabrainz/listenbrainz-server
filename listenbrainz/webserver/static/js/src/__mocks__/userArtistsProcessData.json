[
  {
    "id": "10. The Fray",
    "count": 41
  },
  {
    "id": "9. The Chainsmokers",
    "count": 44
  },
  {
    "id": "8. The Weeknd",
    "count": 52
  },
  {
    "id": "7. OneRepublic",
    "count": 65
  },
  {
    "id": "6. Ellie Goulding",
    "count": 108
  },
  {
    "id": "5. Maroon 5",
    "count": 118
  },
  {
    "id": "4. Imagine Dragons",
    "count": 176
  },
  {
    "id": "3. Coldplay",
    "count": 321
  },
  {
    "id": "2. Lenka",
    "count": 333
  },
  {
    "id": "1. The Local train",
    "count": 385
  }
]
