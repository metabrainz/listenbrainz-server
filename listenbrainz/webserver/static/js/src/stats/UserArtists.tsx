import * as ReactDOM from "react-dom";
import * as React from "react";

import APIService from "../APIService";
import Bar from "./Bar";
import ErrorBoundary from "../ErrorBoundary";

export type UserArtistsData = Array<{
  id: string;
  count: number;
}>;

export type UserArtistsProps = {
  user: ListenBrainzUser;
  apiUrl: string;
};

export type UserArtistsState = {
  data: UserArtistsData;
};

export default class UserArtists extends React.Component<
  UserArtistsProps,
  UserArtistsState
> {
  APIService: APIService;

  ROWS_PER_PAGE = 25; // Number of atists to be shown on each page
  maxListens = 0; // Number of listens for first artist used to scale the graph
  totalPages = 0; // Total namber of pages, useful for pagination
  currPage = 1; // Current page number

  constructor(props: UserArtistsProps) {
    super(props);

    this.APIService = new APIService(
      props.apiUrl || `${window.location.origin}/1`
    ); // Used to access LB API

    this.state = {
      data: [],
    };
  }

  async componentDidMount() {
    const { user } = this.props;

    // Fetch page number from URL
    const url = new URL(window.location.href);
    if (url.searchParams.get("page")) {
      this.currPage = Number(url.searchParams.get("page"));
    } else {
      this.currPage = 1;
    }

    try {
      const data = await this.APIService.getUserStats(
        user.name,
        undefined,
        undefined,
        1
      );
      this.maxListens = data.payload.artists[0].listen_count;
      this.totalPages = Math.ceil(
        data.payload.total_artist_count / this.ROWS_PER_PAGE
      );
    } catch (error) {
      this.handleError(error);
    }
    this.handlePageChange(this.currPage);
  }

  getData = async (
    userName: string,
    offset: number
  ): Promise<UserArtistsResponse> => {
    const data = await this.APIService.getUserStats(
      userName,
      undefined,
      offset,
      this.ROWS_PER_PAGE
    );
    return data;
  };

  processData = (
    data: UserArtistsResponse,
    offset: number
  ): UserArtistsData => {
    const result = data.payload.artists
      .map((elem, idx: number) => {
        return {
          id: `${offset + idx + 1}. ${elem.artist_name}`,
          count: elem.listen_count,
        };
      })
      .reverse();
    return result;
  };

  handlePageChange = async (page: number): Promise<void> => {
    const { user } = this.props;
    const offset = (page - 1) * this.ROWS_PER_PAGE;
    try {
      const data = await this.getData(user.name, offset);
      this.currPage = page;
      this.setState({ data: this.processData(data, offset) });
    } catch (error) {
      this.handleError(error);
    }
  };

  handleError = (error: Error): void => {
    // Error Boundaries don't catch errors in async code.
    // Throwing an error in setState fixes this.
    // This is a hacky solution but should be fixed with upcoming concurrent mode in React.
    this.setState(() => {
      throw error;
    });
  };

  render() {
    const { data } = this.state;
    const prevPage = this.currPage - 1;
    const nextPage = this.currPage + 1;

    return (
      <div>
        <div className="row">
          <div className="col-md-12" style={{ height: "75em" }}>
            <Bar data={data} maxValue={this.maxListens} />
          </div>
        </div>
        <div className="row">
          <div className="col-md-12">
            <ul className="pager">
              <li className={`previous ${!(prevPage > 0) ? "hidden" : ""}`}>
                {/* eslint-disable-next-line */}
                <a onClick={() => this.handlePageChange(prevPage)}>
                  &larr; Previous
                </a>
              </li>
              <li
                className={`next ${
                  !(nextPage <= this.totalPages) ? "hidden" : ""
                }`}
              >
                {/* eslint-disable-next-line */}
                <a onClick={() => this.handlePageChange(nextPage)}>
                  Next &rarr;
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    );
  }
}

document.addEventListener("DOMContentLoaded", () => {
  const domContainer = document.querySelector("#react-container");
  const propsElement = document.getElementById("react-props");
  let reactProps;
  try {
    reactProps = JSON.parse(propsElement!.innerHTML);
  } catch (err) {
    // Show error to the user and ask to reload page
  }
  const { user, api_url: apiUrl } = reactProps;
  ReactDOM.render(
    <ErrorBoundary>
      <UserArtists apiUrl={apiUrl} user={user} />
    </ErrorBoundary>,
    domContainer
  );
});
