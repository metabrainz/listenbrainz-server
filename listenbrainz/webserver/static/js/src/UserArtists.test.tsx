import * as React from "react";
import { mount, shallow } from "enzyme";

import UserArtists from "./UserArtists";
import * as userArtistsResponse from "./__mocks__/userArtists.json";
import * as userArtistsProcessDataOutput from "./__mocks__/userArtistsProcessData.json";

const props = {
  user: {
    id: 0,
    name: "dummyUser",
  },
  apiUrl: "apiUrl",
};

describe("User Artists Page", () => {
  it("renders", () => {
    // We don't need to call componentDidMount during "mount" because we are
    // passing the data manually, so mock the implementation once.
    jest
      .spyOn(UserArtists.prototype, "componentDidMount")
      .mockImplementationOnce((): any => {});

    const wrapper = mount<UserArtists>(<UserArtists {...props} />);
    const instance = wrapper.instance();

    // eslint-disable-next-line dot-notation
    instance["maxListens"] = 385;
    wrapper.setState({ data: userArtistsProcessDataOutput as any });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });
});

describe("componentDidMount", () => {
  it("set current page to 1 if not provided in URL", () => {
    const wrapper = shallow<UserArtists>(<UserArtists {...props} />);
    const instance = wrapper.instance();

    delete window.location;
    window.location = {
      href: "https://foobar/user/bazfoo/artists",
    } as any;
    instance.componentDidMount();

    // eslint-disable-next-line dot-notation
    expect(instance["currPage"]).toBe(1);
  });

  it("extracts current page from the URL if present", () => {
    const wrapper = shallow<UserArtists>(<UserArtists {...props} />);
    const instance = wrapper.instance();

    delete window.location;
    window.location = {
      href: "https://foobar/user/bazfoo/artists?page=3",
    } as any;
    instance.componentDidMount();

    // eslint-disable-next-line dot-notation
    expect(instance["currPage"]).toBe(3);
  });

  it("sets maxListens and totalPages correctly", async () => {
    const wrapper = shallow<UserArtists>(<UserArtists {...props} />);
    const instance = wrapper.instance();

    // eslint-disable-next-line dot-notation
    const spy = jest.spyOn(instance["APIService"], "getUserStats");
    spy.mockImplementation(() => {
      return Promise.resolve(userArtistsResponse);
    });

    await instance.componentDidMount();
    expect(spy).toHaveBeenCalledWith("dummyUser", undefined, undefined, 1);
    // eslint-disable-next-line dot-notation
    expect(instance["maxListens"]).toBe(385);
    // eslint-disable-next-line dot-notation
    expect(instance["totalPages"]).toBe(7);
  });

  it("calls processData and sets state correctly", async () => {
    const wrapper = shallow<UserArtists>(<UserArtists {...props} />);
    const instance = wrapper.instance();

    // eslint-disable-next-line dot-notation
    const spy = jest.spyOn(instance["APIService"], "getUserStats");
    spy.mockImplementation(() => {
      return Promise.resolve(userArtistsResponse);
    });
    instance.processData = jest.fn().mockImplementationOnce(() => {
      return userArtistsProcessDataOutput;
    });
    delete window.location;
    window.location = {
      href: "https://foobar/user/bazfoo/artists",
    } as any;
    await instance.componentDidMount();

    expect(spy).toHaveBeenCalledWith("dummyUser", undefined, 0, 25);
    expect(instance.processData).toHaveBeenCalledWith(userArtistsResponse, 0);
    expect(wrapper.state("data")).toEqual(userArtistsProcessDataOutput);
  });
});

describe("processData", () => {
  it("processes data correctly", () => {
    const wrapper = shallow<UserArtists>(<UserArtists {...props} />);
    const instance = wrapper.instance();

    expect(instance.processData(userArtistsResponse, 0)).toEqual(
      userArtistsProcessDataOutput
    );
  });
});
