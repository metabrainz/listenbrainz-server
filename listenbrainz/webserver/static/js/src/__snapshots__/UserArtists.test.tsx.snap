// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Artists Page renders 1`] = `
<UserArtists
  apiUrl="apiUrl"
  user={
    Object {
      "id": 0,
      "name": "dummyUser",
    }
  }
>
  <div>
    <div
      className="row"
    >
      <div
        className="col-md-8"
        style={
          Object {
            "height": "75em",
          }
        }
      >
        <Bar
          data={
            Array [
              Object {
                "count": 41,
                "id": "The Fray",
              },
              Object {
                "count": 44,
                "id": "The Chainsmokers",
              },
              Object {
                "count": 52,
                "id": "The Weeknd",
              },
              Object {
                "count": 65,
                "id": "OneRepublic",
              },
              Object {
                "count": 108,
                "id": "Ellie Goulding",
              },
              Object {
                "count": 118,
                "id": "Maroon 5",
              },
              Object {
                "count": 176,
                "id": "Imagine Dragons",
              },
              Object {
                "count": 321,
                "id": "Coldplay",
              },
              Object {
                "count": 333,
                "id": "Lenka",
              },
              Object {
                "count": 385,
                "id": "The Local train",
              },
            ]
          }
          maxValue={385}
        >
          <ResponsiveBar
            axisLeft={
              Object {
                "renderTick": [Function],
              }
            }
            colors="#FD8D3C"
            data={
              Array [
                Object {
                  "count": 41,
                  "id": "The Fray",
                },
                Object {
                  "count": 44,
                  "id": "The Chainsmokers",
                },
                Object {
                  "count": 52,
                  "id": "The Weeknd",
                },
                Object {
                  "count": 65,
                  "id": "OneRepublic",
                },
                Object {
                  "count": 108,
                  "id": "Ellie Goulding",
                },
                Object {
                  "count": 118,
                  "id": "Maroon 5",
                },
                Object {
                  "count": 176,
                  "id": "Imagine Dragons",
                },
                Object {
                  "count": 321,
                  "id": "Coldplay",
                },
                Object {
                  "count": 333,
                  "id": "Lenka",
                },
                Object {
                  "count": 385,
                  "id": "The Local train",
                },
              ]
            }
            enableGridY={false}
            indexBy="id"
            keys={
              Array [
                "count",
              ]
            }
            labelFormat={[Function]}
            labelSkipWidth={0}
            layout="horizontal"
            margin={
              Object {
                "left": 204.8,
              }
            }
            maxValue={385}
            padding={0.15}
            theme={
              Object {
                "axis": Object {
                  "ticks": Object {
                    "text": Object {
                      "fontSize": "14px",
                    },
                  },
                },
                "labels": Object {
                  "text": Object {
                    "fontSize": "14px",
                  },
                },
              }
            }
            tooltip={[Function]}
          >
            <ResponsiveWrapper>
              <Measure
                bounds={true}
                onResize={[Function]}
              >
                <Component
                  bounds={true}
                  contentRect={
                    Object {
                      "bounds": Object {},
                      "client": Object {},
                      "entry": Object {},
                      "margin": Object {},
                      "offset": Object {},
                      "scroll": Object {},
                    }
                  }
                  measure={[Function]}
                  measureRef={[Function]}
                >
                  <div
                    style={
                      Object {
                        "height": "100%",
                        "width": "100%",
                      }
                    }
                  />
                </Component>
              </Measure>
            </ResponsiveWrapper>
          </ResponsiveBar>
        </Bar>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-8"
      >
        <ul
          className="pager"
        >
          <li
            className="previous hidden"
          >
            <a
              href="http://localhost/?page=0"
            >
              ← Previous
            </a>
          </li>
          <li
            className="next hidden"
          >
            <a
              href="http://localhost/?page=2"
            >
              Next →
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</UserArtists>
`;
